File: src/main/java/leetcode/hot100/LeetCode46.java
package leetcode.hot100;

import java.util.ArrayList;
import java.util.List;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 18:43
 */
public class LeetCode46 {

    public List<List<Integer>> permute(int[] nums) {
        if (nums == null || nums.length == 0) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();
        dfs(nums, res, 0);
        return res;
    }

    private void dfs(int[] nums, List<List<Integer>> res, int index) {
        if (index == nums.length) {
            res.add(toList(nums));
            return;
        }
        for (int i = index; i < nums.length; i++) {
            swap(nums, i, index);
            dfs(nums, res, index + 1);
            swap(nums, i, index);
        }
    }

    private List<Integer> toList(int[] nums) {
        List<Integer> res = new ArrayList<>();
        for (int num : nums) {
            res.add(num);
        }
        return res;
    }

    private void swap(int[] nums, int i, int index) {
        int temp = nums[i];
        nums[i] = nums[index];
        nums[index] = temp;
    }
}

File: src/main/GO/leetcode/lists/hot100/LeetCode46.go
package hot100

func permute(nums []int) [][]int {
	if len(nums) == 0 {
		return [][]int{}
	}
	list := make([][]int, 0)
	permuteHelper(nums, 0, &list)
	return list
}

func permuteHelper(nums []int, index int, res *[][]int) {
	if len(nums) == index {
		//go study
		c := append([]int(nil), nums...)
		*res = append(*res, c)
		return
	}
	for i := index; i < len(nums); i++ {
		swap(nums, i, index)
		permuteHelper(nums, index+1, res)
		swap(nums, i, index)
	}
}

func swap(nums []int, i int, index int) {
	nums[i], nums[index] = nums[index], nums[i]
}
